### ISSUE TITLE
Shouldn't get be available as a function?

### ISSUE THREAD

**Issue by rcarmo** [hylang/hy#683]
Here's a use case for this:

``` hy
=> (setv a {"foo" {"bar" {"baz" 3}}})
=> a
{u'foo': {u'bar': {u'baz': 3L}}}
=> (reduce get ["foo" "bar" "baz"] a)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
NameError: name 'get' is not defined
=> (reduce (fn [x y] (get x y)) ["foo" "bar" "baz"] a)
3L
```

**Comment by rcarmo** [comment]
*bump*

This would be really handy in a couple more instances (I've already turned that reduce with a lambda into a macro for my own use, but...)

**Comment by paultag** [comment]
Sure. Let's PR that into shadow

**Comment by zackmdavis** [comment]
@rcarmo 
> reduce with a lambda into a macro for my own use, but

It looks like variadic `get` already behaves the way you want? 
```
$ hy
hy 0.10.1 using CPython(default) 3.4.1 on Linux
=> (setv a {"foo" {"bar" {"baz" 3}}})
=> (get a "foo" "bar" "baz")
3
```

**Comment by rcarmo** [comment]
It does? Wow. Had no idea, never even thought of trying. 

Is this git tip or release? I was kind of hoping for an updated release...

**Comment by zackmdavis** [comment]
Release. Git blame (git praise?) [suggests](https://github.com/hylang/hy/blame/854940ca/hy/compiler.py#L1184-1190) it was [introduced in](https://github.com/hylang/hy/commit/62522a5f8) December 2013 before the 0.9.12 release.

**Comment by gilch** [comment]
Is this working already? Can we close this?

**Comment by tuturto** [comment]
```get``` isn't available as a function:

```
hy 0.12.1+23.g92d5d6b using CPython(default) 3.4.3 on Linux
=> (reduce get ["foo" "bar" "baz"] a)
Traceback (most recent call last):
  File "/home/tuukka/programming/python/projects/hy/hy/importer.py", line 146, in hy_eval
    return eval(ast_compile(expr, "<eval>", "eval"), namespace)
  File "<eval>", line 1, in <module>
NameError: name 'get' is not defined
```

**Comment by Kodiologist** [comment]
Yeah, I was thinking of doing a PR for this after #1261 is in.

**Comment by ekaschalk** [comment]
I noticed `apply` is also not shadowed and didn't see any reference to it in #1261.

I have several use-cases. One such is hypothesis which exposes `.map` as a way to apply map to a strategy, but exposes no `.*map`.

So with a reader macro curry as `#$` I cannot do:
```
(->> x
  (.map #$(apply (fn [x y]))
```

And so I instead shadow apply myself.

Is there any reasoning not to have all such hy functionality shadowed? Is there any list of hy api components not availible as functions?

**Comment by Kodiologist** [comment]
Can you describe your use case in more detail, or provide a simpler one? It's confusing to me that you'd want to shadow `apply`, and I don't know anything about Hypothesis.

> Is there any reasoning not to have all such hy functionality shadowed?

Yes, some special forms couldn't reasonably be shadowed, such as `setv`.

> Is there any list of hy api components not availible as functions?

No. If you want to make such a list, you can compare `hy.core.shadow` to the `@builds(...)` decorators in `hy.compiler`, which define special forms, and the macro definitions in `hy.core.bootstrap` and `hy.core.macros`.

**Comment by ekaschalk** [comment]
It's useful when you have a higher-ordered function that you want it to apply the passed function variadically (like *map).

None of the examples are simple but the use-case I have is similar to:

```
(->> [[[x1 x2 ...] [[y11 y12 ...] [y21 y22 ..]]]
         ....]

   ;; dont have access to *map
   (.map #$(apply (fn [x unpack] (+ x (reduce + unpack)))

   (= [[x1 x2 y11 y12 y21 y22 ...]
        ...])
```

And then another I have a function that applies a function to a list. I want to merge the lists but `merge-with` is variadic so I would do `(func #$(apply merge-with))`

**Comment by gilch** [comment]
I do think that special forms and macros should be clearly marked as such in the documentation. Shadowed forms should be labeled both.

**Comment by Kodiologist** [comment]
@ekaschalk I still don't understand your example, perhaps because I'm unfamiliar with Hypothesis. In particular, what object is the sharp macro `$` producing that evidently has a method named `map`?

**Comment by ekaschalk** [comment]
My apologies it should be `->` thread first.

Hypothesis generates data with the following shape. So it isn't actually a list, it is a function that generates a list with that shape. So to modify the generated list you can't map over it as it is a function, not a list.

So this generator exposes `.map` for updating the generator, but no `.*map` for variadically updating the generator.

**Comment by Kodiologist** [comment]
Okay, if I understand correctly, you're trying to do the following:

    (=
      (.map obj (fn [*args]
        (apply (fn [x unpack] (+ x (reduce + unpack))) args)))
      obj2)

If I'm not mistaken, that could be written more concisely as:

    (=
      (.map obj (fn [x unpack]
        (+ x (reduce + unpack)))
      obj2))

Or, yet more concisely,

    (=
      (.map obj (fn [x unpack]
        (sum unpack [x])))
      obj2)

**Comment by ekaschalk** [comment]
Hopefully this illustrates more clearly:

```
(tuple (map (fn [x] (sum (get x 1) (get x 0)))
            [[[1 2] [[1 2 3] [1 2 3 4]]]
             [[2 3] [[2 3 4] [2 3 4 5]]]]))
>>> ([1, 2, 1, 2, 3, 1, 2, 3, 4], [2, 3, 2, 3, 4, 2, 3, 4, 5])

(tuple (*map (fn [x unpack] (sum unpack x))
            [[[1 2] [[1 2 3] [1 2 3 4]]]
             [[2 3] [[2 3 4] [2 3 4 5]]]]))
>>> ([1, 2, 1, 2, 3, 1, 2, 3, 4], [2, 3, 2, 3, 4, 2, 3, 4, 5])

(tuple (map (fn [x unpack] (sum unpack [x]))
            [[[1 2] [[1 2 3] [1 2 3 4]]]
             [[2 3] [[2 3 4] [2 3 4 5]]]]))
>>> TypeError: _hy_anon_fn_1() missing 1 required positional argument: 'unpack'

(tuple #$(apply map (fn [x unpack] (sum unpack [x])))
            [[[1 2] [[1 2 3] [1 2 3 4]]]
             [[2 3] [[2 3 4] [2 3 4 5]]]]))
>>> name apply not found

(tuple #$(shadowed_apply map (fn [x unpack] (sum unpack [x])))
            [[[1 2] [[1 2 3] [1 2 3 4]]]
             [[2 3] [[2 3 4] [2 3 4 5]]]]))
>>> ([1, 2, 1, 2, 3, 1, 2, 3, 4], [2, 3, 2, 3, 4, 2, 3, 4, 5])
```

Need `*map` to apply it otherwise need to get the components.

**Comment by Kodiologist** [comment]
Try using parameter unpacking:

    => (tuple (map (fn [[x unpack]] (sum unpack x))
    ...             [[[1 2] [[1 2 3] [1 2 3 4]]]
    ...              [[2 3] [[2 3 4] [2 3 4 5]]]]))
    ([1, 2, 1, 2, 3, 1, 2, 3, 4], [2, 3, 2, 3, 4, 2, 3, 4, 5])

**Comment by ekaschalk** [comment]
Oh that's a very nice feature of hy, thanks! Knew it did matching in assignments (like python) but not parameters (unlike python). Is this documented anywhere, I don't see it in docs for `defn` or through a search.

**Comment by Kodiologist** [comment]
> Is this documented anywhere

I don't know. To be honest, I hadn't realized the feature was removed from Python in Python 3 ([PEP 3113](https://www.python.org/dev/peps/pep-3113/)) and that Hy does the destructuring itself in `compile_function_def`.

**Comment by gilch** [comment]
@ekaschalk The shadowed `apply` thing is off topic, and probably belonged in #647. It may be moot since the unpacking operators replaced `apply` in #1325. I'm also not sure if removing `apply` was a good idea or not, but if you have more to say about it, open a new issue.

The probem with shadowing Hy's `get` is that it does not correspond to only one operator in Python, but three: `getitem`, `setitem`, and `delitem`.
```
=> (def spam {})
spam = {}
None
=> (setv (get spam "a") "A")
spam['a'] = 'A'
None
=> (get spam "a")
spam['a']
'A'
=> (del (get spam "a"))
del spam['a']
None
```
Maybe we should just add these there operators to core.

**Comment by Kodiologist** [comment]
I would expect shadowed `get` to work like `getitem`. The obvious implementation is something like

    (defn get [thing &rest keys]
      (for [k keys]
        (setv thing (get thing k)))
      thing)

**Comment by gilch** [comment]
No, the obvious implementation is `from operator import getitem as get`. 

Your implementation does more than `getitem`. It's more like Clojure's `get-in`, but without the optional default.
```Clojure
user=> (get-in [:a [:b [:c]]] [1 1 0] :absent)
:c
user=> (get-in [:a] [1 1 0] :absent)
:absent
```
I'm not opposed to adding something like Clojure's `get-in` to Hy, but Toolz already has one #1339.

**Comment by Kodiologist** [comment]
> No, the obvious implementation is `from operator import getitem as get`.

This would fail to emulate `get` with more than two arguments.

**Comment by gilch** [comment]
That is a good point. I forgot our special form can take multiple keys.

I'd like to have the defaults of Clojure's `get-in`, but that's no reason not to shadow `get`. Clojure's `get` does take a default like `get-in`, which is different than Hy. I thought about changing it, but Hy's version of the special form lets us use it like a [Common Lisp *place*](http://www.gigamonkeys.com/book/variables.html), which makes more sense for Python's mutable data structures. If we added a the default logic to the compilation of the special form, it would be difficult to use it in `setv` or `del` statement.

I think generalized places for use in a `setv` or `del` could be done in a macro by using the `__setitem__`/`__delitem__` magic in a custom place class, but that's probably not what we want by default for a simple `get`.
